core/include/vecmem/containers/impl/device_vector.ipp

```diff
@@ -225,6 +227,50 @@ VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::push_back(
     return index;
 }
 
+template <typename TYPE>
+VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::reserve(size_type n)
+    -> size_type {
+    // This can only be done on a resizable vector.
+    assert(m_size != nullptr);
+
+#if defined(__cpp_lib_is_implicit_lifetime) && \
+    __cpp_lib_is_implicit_lifetime >= 202302L
+    static_assert(std::is_implicit_lifetyime_v<TYPE>,
+                  "Type `T` in `device_vector<T>::reserve` is not an implicit "
+                  "lifetype type (according to the C++23-compliant "
+                  "definition), so slots cannot be safely reserved.");
+#else
+    static_assert(details::is_implicit_lifetime_v<TYPE>,
+                  "Type `T` in `device_vector<T>::reserve` is not an implicit "
+                  "lifetype type (according to a non-C++23-compliant "
+                  "definition), so slots cannot be safely reserved.");
+#endif
+
+    device_atomic_ref<size_type> asize(*m_size);
+    const size_type index = asize.fetch_add(n);
+    assert((index + n - 1) < m_capacity);
```

Comment:
Why not

```suggestion
    assert((index + n) <= m_capacity);
```

? The compiler may have a slightly easier time with that one...

Reply:
That's fair!